### PR TITLE
Reduce the need for creating new TableEntity subclasses when working with Table Storage.

### PR DESCRIPTION
**Pull request by antoinne85**
# **This PR is not ready to merge.**

**I only submitted it to get a feel for whether or not the project maintainers are interested in this potential contribution before I spend hours writing tests, running tests, and doing all of the additional work to make sure the change is truly ready to be merged.**

If this kind of addition seems desirable to the project maintainers I'll continue forward with writing additional tests and such.

Now that that's out of the way, read on.

I've been working with table storage for a while now, and have found it somewhat labor intensive to keep from tightly coupling data objects from the Azure.Storage libraries. I find myself creating TableEntity subclasses that are almost 1-1 duplicates of data objects I want to work with and then mapping to-and-from these objects before they leave repositories that interact with Table Storage (through a thin, reusable abstraction).

Lately I've been thinking about how one might get around the need to do this while not leaking the fact that Table Storage is in use under-the-covers and reducing the amount of boilerplate one has to write to achieve that end goal.

Eventually I started examining the source in this repository, pulled it down, and began poking at it. It quickly became clear that most of the work to achieve my goal had already been written. _TableEntity_ had ReadEntity and WriteEntity methods that more-or-less fully de/rehydrate an arbitrary object via reflection (_with a caching mechanism, even_). 

It seemed like a fairly small step to have it work against arbitrary objects--thereby allowing the consumer to avoid creating objects that extend TableEntity and instead use their own data objects and a reusable form of TableEntity to interact with table storage.

**It's important to note that I haven't spent a great deal of time with this source yet and this idea is perhaps not fully baked yet,** but before I spent hours upon hours getting up to speed on all the relevant bits I wanted to get some feedback on whether or not an addition like this is desirable. Any and all feedback is welcome.

More notes in the source below.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@antoinne85__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

File: Lib/Common/Table/TableEntity.cs [hunk 1]
```diff
@@ -108,19 +108,26 @@ public TableEntity(string partitionKey, string rowKey)
         /// <param name="operationContext">An <see cref="OperationContext"/> object that represents the context for the current operation.</param>
         public virtual void ReadEntity(IDictionary<string, EntityProperty> properties, OperationContext operationContext)
```
Comment (antoinne85) [inline Lib/Common/Table/TableEntity.cs]: Previously, this method contained all of the logic to read an entity. It seemed already prepared to handle objects of arbitrary types, but "this" was always passed into the reflection-y methods, effectively forcing it to always work against the _TableEntity_ type (or whatever subclass the user had created).

I simply pulled this code out into another method that allows you to pass in the object you want to use in place of "this" and called that method from here with "this" to keep the existing functionality intact.

**Comment by antoinne85** [comment]
Example:

Suppose I have the following data object:
```csharp
public class Person
{
    public string State { get; set; }
    public string SocialSecurityNumber { get; set; }
    public int SomeOtherProperty { get; set; }
}
```

In order to start storing this object in table storage, I either have to introduce TableEntity as base class for my Person class (and thereby start leaking the Azure dependencies around other parts of my code) or I have to create a new class that looks like

```csharp
public class PersonEntity : TableEntity
{
    public string State { get; set; }
    public string SocialSecurityNumber { get; set; }
    public int SomeOtherProperty { get; set; }
}
```
as well as map the PartitionKey and RowKey properties introduced by _TableEntity_ to the properties I want to use index my PersonEntity class. Further, I then have to produce some code to map back and forth between my Person class and PersonEntity class to keep my Azure dependencies from leaking outside of whatever storage abstraction I've created.

The hope is that I can now keep the Azure dependencies from leaking with less work by, wherever I need to convert my object to an ITableEntity, I simply do something like:

```csharp
var entities = myPeople.Select(p => new GenericTableEntity<Person>(p, p.State, p.SocialSecurityNumber));
```

Now I have TableEntities ready to use with the Table Storage APIs without having to write a bunch of boilerplate.

**Comment by azurecla** [comment]
__@antoinne85__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

**Comment by erezvani1529** [comment]
Hi @antoinne85,

Thank you for your contribution and apologies about the delay in response.

We are definitely interested in the idea of your PR. With that said, seems like it would be best for this feature to be available for more complex user provided types and therefore, we recommend you consider using the [Flatten](https://github.com/Azure/azure-storage-net/blob/master/Lib/Common/Table/TableEntity.cs#L334) and [ConvertBack](https://github.com/Azure/azure-storage-net/blob/master/Lib/Common/Table/TableEntity.cs#L153) methods provided on TableEntity instead of the  Read/WriteEntity methods which you have used.

The Flatten/ConvertBack APIs, help serialize more complex entities and in the case of regular POCO entities with no nested or complexly structured properties, provide the exact same behavior as reflection based read and write. 

Please have a look at the suggested change and let us know your opinion so that we can proceed to the next steps with this contribution.

<One more point, please submit your changes against Dev branch instead of Master :)>

Thanks!

**Comment by antoinne85** [comment]
Hey, @erezvani1529, thanks for getting back to me.

Great. Armed with the knowledge that there's some interest I'll continue work on this. Thanks especially for pointing me to Flatten and ConvertBack. It's always a little daunting to pick up a code base once it reaches this size and orient yourself.

I'm going to put some more effort into this and acquaint myself with the tests and I'll pop back in when I have a more complete update.

**Comment by antoinne85** [comment]
One more thing--with the switch to Flatten and ConvertBack, there's even less movement in this change than before.

What kind of tests would you expect to find with a change like this. The core functionality of the library is untouched, there's just a new kind of TableEntity.

I'm guessing I should include (at least) a test to cover the ability for the object to round-trip successfully, but not so thorough as to rehash the same unit tests that are already covering the Flatten and ConvertBack methods.

Thoughts?

**Comment by erezvani1529** [comment]
Hi @antoinne85,
 
Make sure you cover both success and failure paths. For the success path, as you mentioned the end-to-end should be covered(you can make use of the existing test entities).  For failure cases, we would like to identify them if any and make sure we are failing as expected for un-supported scenarios. So, you are right: no need to re-hash all the tests that are already being covered. :)

Thanks,
Elham

**Comment by antoinne85** [comment]
Hey @erezvani1529,

I've finally managed to carve out a few minutes to spend some more time on this.

I've pushed a couple of minor updates. In working on them, I've run across some things that require me to ask more questions.

1. In the one simple test included, when retrieving data using `TablePayloadFormat.JsonNoMetadata` the integer field `Age` comes back as a string and is not properly converted to an integer. I've not used the TablePayloadFormat options before, but based on the documentation this seems to be the expected behavior. Can you confirm?
1. Am I headed down the proper path in terms of testing? Are there additional success-case tests you might like to see?
1. With regards to failure-case testing, the bulk of the change here is simply piggybacking on existing code that is under test. The type constraints (`where T : class, new()`) prevent you from using incompatible types at compile-time and I'm having difficulty imagining other runtime failure scenarios that aren't already covered under the existing test suite. I'm happy to add more coverage, but I'm afraid that my unfamiliarity with the real nuts and bolts of what's in the heart of the library make it difficult for me to speculate on what might be still required.

**Comment by erezvani1529** [comment]
Hi @antoinne85,

Apologies for the delay in getting back. Thanks for putting more effort into this PR. What you have started has been on our radar for a while and @doguva-msft has also been working on it internally and as you can see, his [PR](https://github.com/Azure/azure-storage-net/pull/436) here is almost complete. Since you have started working on this feature sooner, I would like to ask if you would be able to address the comments to use Flatten/ConvertBack and complete up the change sooner or you would be ok with us proceeding with Dogu's change with #436 ? We will take it further based on your answer :)

Thanks again,
Elham

**Comment by antoinne85** [comment]
Hey @erezvani1529. 

I was actually looking at that PR just a few days ago. 

It seems better put-together than what I had done. Its also got a much better name for the class. :)

I'm far more interested in the change going in than I am in "getting credit" or anything like that. 

So, by all means, feel free to continue forward with @doguva-msft's implementation. I look forward to using it in the near future. 

I'm going to go ahead and close this one out to clean up your list.

**Pull request closed without merging**